src/smartpeak/source/core/SampleGroupProcessor.cpp

```diff
@@ -26,10 +26,92 @@
 #include <SmartPeak/core/FeatureMetadata.h>
 #include <SmartPeak/io/InputDataValidation.h>
 #include <OpenMS/FORMAT/FeatureXMLFile.h>
+#include <SmartPeak/io/SelectiveDilutionsParser.h>
 #include <plog/Log.h>
 
 namespace SmartPeak
 {
+  ParameterSet SelectFromDilution::getParameterSchema() const
+  {
+    return ParameterSet();
+  }
+
+  void SelectFromDilution::process(
+    SampleGroupHandler& sampleGroupHandler_IO,
+    const SequenceHandler& sequenceHandler_I,
+    const ParameterSet& params_I,
+    const Filenames& filenames
+  ) const
+  {
+    LOGD << "START SelectFromDilution";
+
+    ParameterSet params(params_I);
+    params.merge(getParameterSchema());
+
+    std::map<std::string, int> selective_dilution_map;
+    try
+    {
+      SelectiveDilutionsParser::read(filenames.selectiveDilutions_csv_i, selective_dilution_map);
+    }
+    catch (const std::exception& e)
+    {
+      LOGE << "Failed to read selective dilutions file [" << filenames.selectiveDilutions_csv_i << "] : " << e.what();
+      return;
+    }
+
+    std::map<std::string, std::set<int>> existing_dilutions_map;
+    for (const std::size_t& index : sampleGroupHandler_IO.getSampleIndices())
+    {
+      auto& injection = sequenceHandler_I.getSequence().at(index);
+      for (const auto& feature : injection.getRawData().getFeatureMap())
+      {
+        for (const auto& sub_feature : feature.getSubordinates())
+        {
+          const auto& native_id = sub_feature.getMetaValue("native_id");
+          existing_dilutions_map.emplace(native_id, std::set<int>());
+          existing_dilutions_map[native_id].insert(injection.getMetaData().dilution_factor);
+        }
+      }
+    }
+
+    OpenMS::FeatureMap new_feature_map;
+    for (const std::size_t& index : sampleGroupHandler_IO.getSampleIndices())
+    {
+      auto& injection = sequenceHandler_I.getSequence().at(index);
+      for (const auto& feature : injection.getRawData().getFeatureMap())
+      {
+        auto new_feature = feature;
+        new_feature.getSubordinates().clear();
+        for (const auto& sub_feature : feature.getSubordinates())
+        {
+          bool add_feature = true;
+          const auto& native_id = sub_feature.getMetaValue("native_id");
+          if (selective_dilution_map.count(native_id))
+          {
+            const auto preferred_dilution = selective_dilution_map.at(native_id);
+            if (injection.getMetaData().dilution_factor == preferred_dilution)
+            {
+              add_feature = true;
+            }
+            else
+            {
+              // if there is no other feature is listed with the preferred dilution, we will use that one
+              const auto& existing_dilutions = existing_dilutions_map[native_id];
+              add_feature = (std::find(existing_dilutions.begin(), existing_dilutions.end(), preferred_dilution) == existing_dilutions.end());
```

Comment:
In most cases this should be a fair assumption.  However, I can also see this potentially leading to problems.  I would recommend adding in a parameter so that the user can control if they want to default to the available dilution if the preferred dilution is not present or simply omit the value.

Reply:
Added one method parameter to choose the behaviour
"preferred" the original behaviour 
"exlclusive" if the component is not found at the right dilution don't add it whatever dilution it is.